### ISSUE TITLE
Add a common type between all AudioStreamPlayer variants

### ISSUE THREAD

**Issue by FlooferLand** [godotengine/godot-proposals#7356]
### Describe the project you are working on

moddable 2D side-scroll zombie survival game with a focus on vehicles

### Describe the problem or limitation you are having in your project

There's no way to refer to an `AudioStreamPlayer` regardless of it's variant _(1D/2D/3D)_. This is particularly annoying in C# and other typed languages, as you want to call methods like `Play()` and use properties like `PitchScale` that all variants have, without wanting to refer to a variant in particular.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Sometimes you might want to change whenever you want your audio to be mono or stereo without refactoring your scripts, and in some cases like working on a C# extension script you have to separate your method and write a bunch of overloads for every player variant; you also have to use `Node.Set/Get` to do anything with the stream player in that case.

The current way the stream players are separated also doesn't follow general node-naming conventions I've seen troughout my time using Godot. Seeing an "AudioStreamPlayer" that has two "AudioStreamPlayer2D/3D" variants automatically makes you assume the 2D/3D variants inherit from the 1D/mono variant.

This change could help users, as well as audio add-on developers.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

There could be an interface called `IAudioStreamPlayer` that all `AudioStreamPlayer`s implement, containing all methods and fields of `AudioStreamPlayer`.

_Additionally, the more complicated approach that i feel is the best one:_ Make the 2D and 3D variants both inherit `AudioStreamPlayer`_(mono variant)_ instead of `Node2D/3D`. This would also allow code to be shared between all 3 variants.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Improvement would be used or aknowledged somewhat often. Would be used a lot by people that often work with audio.

### Is there a reason why this should be core and not an add-on in the asset library?

Making this into an add-on would be evil and a crime againt the human race.

**Comment by AThousandShips** [comment]
There isn't a "1D" variant, it's a non-spatial one

Having this kind of setup would mean manually configuring a bunch of stuff that has to be integrated, which doesn't seem likely with how the .NET code is generated, not manually written for these types

**Comment by KoBeWi** [comment]
AudioStreamPlayers is not mono.

**Comment by AThousandShips** [comment]
> Making this into an add-on would be evil and a crime againt the human race.

Please take these questions seriously

**Comment by AThousandShips** [comment]
> Additionally, the more complicated approach that i feel is the best one: Make the 2D and 3D variants both inherit AudioStreamPlayer(mono variant) instead of Node2D/3D. This would also allow code to be shared between all 3 variants.

This won't work at all, they need to be `Node2D/3D` to have the spatial properties

**Comment by Calinou** [comment]
You can adjust AudioStreamPlayer2D/3D's properties to make them *sound* like a non-positional AudioStreamPlayer.

In AudioStreamPlayer2D, set max distance to a very high value and panning strength to `0.0`.


In AudioStreamPlayer3D, set Attenuation to **Disabled**, panning strength to `0.0` and **Cutoff Hz** to 20500.

**Comment by FlooferLand** [comment]
> > Additionally, the more complicated approach that i feel is the best one: Make the 2D and 3D variants both inherit AudioStreamPlayer(mono variant) instead of Node2D/3D. This would also allow code to be shared between all 3 variants.
> 
> This won't work at all, they need to be `Node2D/3D` to have the spatial properties

Ah i see, so perhaps making a common interface between all 3 would be feasable?
I'm not sure GDScript has support for interfaces, so that might be an issue.

**Comment by FlooferLand** [comment]
> In 4.0 and later, you can adjust AudioStreamPlayer2D/3D's properties to make them _sound_ like a non-positional AudioStreamPlayer.
> 
> In AudioStreamPlayer2D, set max distance to a very high value and panning strength to `0.0`.
> 
> In AudioStreamPlayer3D, set **Attenuation Model** to **Disabled**, panning strength to `0.0` and **Attenuation Filter > Cutoff Hz** to 20500 (this disables the low-pass filter when far away, though it might not be required with the **Disabled** attenuation mode).

Really the main problem I've found is that there's no way to refer to any AudioStreamPlayer without specifically referring to one variant.
I'm not sure an interface would be the best way to handle this, since that still wouldnt let you change the pitch, volume, and other properties, unless the setters/getters for those were made into callable methods again which isn't ideal.
Having some class they all inherit from that just stores shared data _(pitch, volume, stream, play/stop methods, etc)_ sounds like a decent solution.
Really the main problem then would be naming that shared class, without confusing users even further.

**Comment by AThousandShips** [comment]
> Ah i see, so perhaps making a common interface between all 3 would be feasable?

No, as I said above, it wouldn't be as these nodes are automatically generated, not hand coded, so how would this be achieved?

**Comment by FlooferLand** [comment]
> Having some class they all inherit from that just stores shared data _(pitch, volume, stream, play/stop methods, etc)_ sounds like a decent solution.

^ would this work better perhaps?
Not sure how the node generation works, but the class they all share could probably reside in [`scene/audio`](https://github.com/godotengine/godot/tree/master/scene/audio), with the audio nodes sharing that class via multiple inheritance

**Comment by Calinou** [comment]
> I'm not sure an interface would be the best way to handle this, since that still wouldnt let you change the pitch, volume, and other properties, unless the setters/getters for those were made into callable methods again which isn't ideal.
Having some class they all inherit from that just stores shared data (pitch, volume, stream, play/stop methods, etc) sounds like a decent solution.

I believe these properties share names across all 3 nodes, so you can use `.set()` and `.get()` to perform duck typing.

**Comment by FlooferLand** [comment]
> I believe these properties share names across all 3 nodes, so you can use `.set()` and `.get()` to set/get those properties in a duck-typed fashion (or use a `Node` type hint when assigning the node to a variable).

It's what i ended up doing, but in some cases like C# extension methods its impossible to do without making your extension methods available to everything that inherits from `Node`; It's either that, or making 3 overloads for every method. Using set/get is a bit inconvenient as well.

This does seem it wouldnt be as worth implementing as i originally thought it'd be; having to figure out a name for the new class that wouldnt confuse users + the only solution i have in mind being kind of lazily slapped together. I originally made this proposal as i thought fixing my inconvenience while also making the engine's node naming more consistent _and_ allowing people to refer to `AudioStreamPlayer` interchangibly _(as people use `Node` to refer to `Node2D` or `Node3D`)_ would be worth the hassle of refactoring a few tens of lines of code or slapping on an interface, but I've only now realized `AudioStreamPlayer2D` alone is about 500 lines of code. Yikes.

Feel free to close the thread, or perhaps for other people's opinions. Sorry for the time waste!
